### ISSUE TITLE
editor.suggest.insertMode not work together with clangd extension

### ISSUE THREAD

**Issue by michaelwolfram** [clangd/vscode-clangd#665]
Type: <b>Bug</b>

`"editor.suggest.insertMode": "replace",` is in my settings so I have the following expectation:

Expected:
- let's have `abc_test` and cursor after the `e`
- Ctrl+Space
- Accept suggestion and text becomes `abc_text`

Actual:
- let's have `abc_test` and cursor after the `e`
- Ctrl+Space
- Accept suggestion and text becomes `abc_textst`

`Help: Start extension bisect` suggested that clangd is causing this issue. After deactivating clangd the expected behavior occurs.

I hope I provided enough required information.
Thanks!

Extension version: 0.1.29
VS Code version: Code 1.91.1 (f1e16e1e6214d7c44d078b1f0607b2388f29d729, 2024-07-09T22:06:49.809Z)
OS version: Windows_NT x64 10.0.19045
Modes:
Connection to 'ssh-remote+workstation' could not be established

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|13th Gen Intel(R) Core(TM) i7-13850HX (28 x 2304)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|undefined|
|Memory (System)|31.64GB (20.16GB free)|
|Process Argv|--crash-reporter-id 9c116f79-07e6-4068-9cff-8f83537cdfbc|
|Screen Reader|no|
|VM|0%|

Connection to 'ssh-remote+workstation' could not be established
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vspor879:30202332
vspor708:30202333
vspor363:30204092
vscoreces:30445986
vscod805cf:30301675
binariesv615:30325510
vsaa593cf:30376535
py29gd2263:31024239
c4g48928:30535728
azure-dev_surveyone:30548225
a9j8j154:30646983
962ge761:30959799
pythongtdpath:30769146
welcomedialog:30910333
pythonnoceb:30805159
asynctok:30898717
pythonregdiag2:30936856
pythonmypyd1:30879173
2e7ec940:31000449
pythontbext0:30879054
accentitlementst:30995554
dsvsc016:30899300
dsvsc017:30899301
dsvsc018:30899302
cppperfnew:31000557
dsvsc020:30976470
pythonait:31006305
dsvsc021:30996838
bdiig495:31013172
pythoncenvpt:31062603
a69g1124:31058053
dvdeprecation:31068756
dwnewjupytercf:31046870
impr_priority:31102340
nativerepl2:31104044
refactort:31108082
ccplt:31103426
pythonrstrctxt:31103193
wkspc-onlycs-t:31106322

```

</details>

**Comment by HighCommander4** [comment]
> Expected:
> 
>    * let's have `abc_test` and cursor after the `e`
>    * Ctrl+Space
>    * Accept suggestion and text becomes `abc_text`

I'm not quite following this fully. Could you give some more details please?

 * What type of file are you experiencing this issue in -- a C or C++ source file?
 * Can you give an example of the complete file contents (i.e. what else is there besides `abc_test`)?
 * What is the suggestion you're accepting (it sounds like it's `text`), and where is it coming from? Is it coming from clangd?

**Comment by michaelwolfram** [comment]
Hi,
- I am talking about C++ files here.
- Unfortunately, I cannot share the original complete file but this minimal example also fails for me:
   ```
   struct Test
   {
       double abc_test{};
       double abc_text{};
       double abc_something{};
       double abc_else{};
   };
   
   void someFunction()
   {
       Test test{};
       test.abc_test = 42.;
   }
   ```
- Having the cursor after `abc_`, Ctrl+space, accept any suggestion and then I e.g. get `test.abc_elsetest = 42.;` 

Here's further analysis I did with the two extensions
- both extensions off results in `test.abc_something = 42.;`
   ![image](https://github.com/user-attachments/assets/670b7190-a751-41b1-80a7-c94183671e5a)
   ->
   ![image](https://github.com/user-attachments/assets/21669c6a-3beb-4ecd-a7d0-b2b36a759f7e)
- only CPP extensions on results in "No suggestions".
- only clangd extensions on results in `test.abc_somethingelse = 42.;`
   ![image](https://github.com/user-attachments/assets/adf67e18-0ecc-4557-b039-7ffb7ce78185)
   ->
   ![image](https://github.com/user-attachments/assets/fbcc6b2a-adad-4a81-895b-e74eeb9664af)
- both extensions on results in `test.abc_somethingelse = 42.;`
   same image as above

This means the non-working suggestions with the blue cube is coming from clangd I guess. And the working one with "abc" in front from native vscode?

Thanks for your help.

**Comment by HighCommander4** [comment]
Thanks for the added details, I understand the issue now.

The documentation for `"editor.suggest.insertMode"` says:

> Controls whether words are overwritten when accepting completions. Note that this depends on extensions opting into this feature.

It looks like clangd has not yet opted into this feature -- but it could, so let's keep this issue open to track implementing that.

**Comment by HighCommander4** [comment]
> This means the non-working suggestions with the blue cube is coming from clangd I guess. And the working one with "abc" in front from native vscode?

That's right. The native vscode suggestions are purely textual, basically whatever words appear in the file.

To illustrate the difference, let's add a global variable declaration that begins with `abc_`:

```c++
int abc_global;

struct Test
{
    double abc_test{};
    double abc_text{};
    double abc_something{};
    double abc_else{};
};

void someFunction()
{
    Test test{};
    test.abc_test = 42.;
}
```

Now, vscode's native completion will offer `abc_global`, but clangd will not because clangd knows that you're in a context where only the name of a member of `struct Test` is valid.

**Comment by HighCommander4** [comment]
> It looks like clangd has not yet opted into this feature -- but it could, so let's keep this issue open to track implementing that.

The first step here will be to understand how extensions are meant to opt into supporting editor.suggest.insertMode.

Is the actual replacement handled by vscode, if an extension somehow opts into it?

Or is the extension expected to handle it? (In which case, for a Language Server Protocol extension like vscode-clangd, the actual logic probably needs to be on the server side, and the next question is how the client should communicate the preference value to the server.)

**Comment by HighCommander4** [comment]
> The first step here will be to understand how extensions are meant to opt into supporting editor.suggest.insertMode.
> 
> Is the actual replacement handled by vscode, if an extension somehow opts into it?
> 
> Or is the extension expected to handle it? (In which case, for a Language Server Protocol extension like vscode-clangd, the actual logic probably needs to be on the server side, and the next question is how the client should communicate the preference value to the server over the protocol.)

I did a bit of research into this. It's mostly the latter case (the server needs to handle it), but there's a bit of a twist.

From the description of `CompletionItem.textEdit` in the [LSP spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItem):

```
 * Most editors support two different operations when accepting a completion
 * item. One is to insert a completion text and the other is to replace an
 * existing text with a completion text. Since this can usually not be
 * predetermined by a server it can report both ranges. Clients need to
 * signal support for `InsertReplaceEdit`s via the
 * `textDocument.completion.completionItem.insertReplaceSupport` client
 * capability property.
```

So, rather than the client communicating the insert mode to the server, the server needs to send the replacement ranges that would apply for both insert modes.

The reason for this, it turns out, is that clients may support letting the user choose which insert mode will be used "on the spot", using different keyboard shortcuts. For vscode, accepting a completion with <kbd>Enter</kbd> will use the default mode (e.g. `"replace"` if you have the default configured to that), but you can also accept a completion with <kbd>Shift</kbd> + <kbd>Enter</kbd> and it will use the other mode. This means that at the time the completion is computed by the server, the insert mode is not yet known.

**Comment by HighCommander4** [comment]
So, to summarize, to support this:

 * The clangd server needs to implement the `InsertReplaceEdit` feature added in LSP 3.16, and use it (send an `InsertReplaceEdit` for `CompletionItem.textEdit`) for clients that indicate `insertReplaceSupport` in the client capabilities
 * No changes to the client are needed

I will accordingly move this issue to the server repo.
